client/state/state_database.go

```diff
@@ -234,8 +234,8 @@ func (s *BoltStateDB) getAllAllocations(tx *boltdd.Tx) ([]*structs.Allocation, m
 }
 
 // PutAllocation stores an allocation or returns an error.
-func (s *BoltStateDB) PutAllocation(alloc *structs.Allocation) error {
-	return s.db.Update(func(tx *boltdd.Tx) error {
+func (s *BoltStateDB) PutAllocation(alloc *structs.Allocation, inBatch bool) error {
+	return s.updateOrBatch(inBatch, func(tx *boltdd.Tx) error {
```

Comment:
Rather than having the boolean behavior flag and this function indirection, let's make the signature really super clear by splitting out a `PutAllocation` and a `PutAllocationBatch`, both of which call into a `putAllocationImpl` with the appropriate boltdb function.

Same for `PutNetworkStatus` and `DeleteAllocationBucket`

Reply:
That makes sense. The naming here is tricky; I'd definitely expect `PutAllocationBatch` to have a slice of allocations as the argument, not a single allocation.

That makes me think - if we add a new method to the interface anyway, maybe we can make it an actual "batch" variant that explicitly saves multiple allocations and not rely on hidden state using boltdb's Batch method? Wdyt?
This would require changing `PersistState` to `PersistStateBatch` too, but that shouldn't be hard to do.

Reply:
> maybe we can make it an actual "batch" variant that explicitly saves multiple allocations 

I'm open to the idea of this but given how we run `PersistState` by iterating over allocation runners (which requires setting and releasing locks on those runner), I suspect that's not going to be very nice for the callers. Give it a go if you'd like though!

Reply:
Yeah handling `destroyedLock` and `stateLock` might be a problem with real batching.

I've fleshed out the two-method approach you suggested and it doesn't look tidy enough to my taste, so wanted to circle back for advice. Conceptually, that boolean flag is an adjustment in behavior (basically opt-in to a perf optimization), not a new behavior, so it looks weird when I have to duplicate more or less the same code in several places. It looks like this:

```
	// PutAllocation stores an allocation or returns an error if it could
	// not be stored.
	PutAllocation(*structs.Allocation) error

	// PutAllocationInBatch has exactly the same semantics as PutAllocation above,
	// with the addition that the write transactions can be shared with other concurrent
	// writes, increasing performance in batch scenarios.
	PutAllocationInBatch(*structs.Allocation) error

	// Get/Put NetworkStatus get and put the allocation's network
	// status. It may be nil.
	GetNetworkStatus(allocID string) (*structs.AllocNetworkStatus, error)
	PutNetworkStatus(allocID string, ns *structs.AllocNetworkStatus) error

	// PutNetworkStatusInBatch has exactly the same semantics as PutNetworkStatus above,
	// with the addition that the write transactions can be shared with other concurrent
	// writes, increasing performance in batch scenarios.
	PutNetworkStatusInBatch(*structs.Allocation) error
```

So while I agree that using booleans to change behavior is a red flag, in this particular case it's arguably not changing semantics, just performance in particular scenarios.

I saw another approach to implementing such behavior adjustments using variadic options, something like this:
```
PutAllocation(*structs.Allocation, ...WriteOption) error
```
(see e.g. https://medium.com/soon-london/variadic-configuration-functions-in-go-8cef1c97ce99  for details)

That would avoid polluting cases where no such option is needed, while allowing these adjustments where needed. What do you think?

Reply:
> That would avoid polluting cases where no such option is needed, while allowing these adjustments where needed. What do you think?

Oh, that's super nice. Go for it.

Reply:
Ok I think it's ready for review.